components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/AutofillCreditCardsAddressesStorage.kt

```diff
@@ -0,0 +1,130 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.storage.sync
+
+import android.content.Context
+import androidx.annotation.GuardedBy
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.withContext
+import mozilla.components.concept.storage.Address
+import mozilla.components.concept.storage.CreditCard
+import mozilla.components.concept.storage.CreditCardsAddressesStorage
+import mozilla.components.concept.storage.UpdatableAddressFields
+import mozilla.components.concept.storage.UpdatableCreditCardFields
+import java.io.Closeable
+import org.mozilla.appservices.autofill.Store as RustAutofillStorage
+
+const val AUTOFILL_DB_NAME = "autofill.sqlite"
+
+/**
+ * An implementation of [CreditCardsAddressesStorage] back by the application-services' `autofill`
+ * library.
+ */
+@SuppressWarnings("TooManyFunctions")
+class AutofillCreditCardsAddressesStorage(
+    context: Context
+) : CreditCardsAddressesStorage, AutoCloseable {
+
+    private val coroutineContext by lazy { Dispatchers.IO }
+
+    private val conn by lazy {
+        AutofillStorageConnection.init(dbPath = context.getDatabasePath(AUTOFILL_DB_NAME).absolutePath)
+        AutofillStorageConnection
+    }
+
+    init {
+        // Set the name of the native library so that we use
+        // the appservices megazord for compiled code.
```

Comment:
Link to https://github.com/mozilla/application-services/issues/3874 to not forget to remove this.